File: code/unclassified/spiral_printing/spiral_print.go
package main

import "fmt"

// Part of Cosmos by OpenGenus Foundation
func PrintSpiral(list [][]int, rows, cols int) {
	var T, B, L, R, dir int = 0, rows - 1, 0, cols - 1, 0

	for {
		if T >= B || L >= R {
			break
		}
		// 0 - traverse right (going Left to Right)
		if dir == 0 {
			for i := 0; i <= R; i++ {
				fmt.Println(list[T][i])
			}
			T++
			dir = 1
		}
		// 1 - traverse down (going Top to Bottom)
		if dir == 1 {
			for i := T; i <= B; i++ {
				fmt.Println(list[i][R])
			}
			R--
			dir = 2
		}
		// 2 - traverse left
		if dir == 2 {
			for i := R; i >= L; i-- {
				//fmt.Println("---> ", R, i)
				fmt.Println(list[B][i])
			}
			B--
			dir = 3
		}
		// 3 - traverse up
		if dir == 3 {
			for i := B; i >= T; i-- {
				fmt.Println(list[i][L])
			}
			L++
			dir = 0
		}
	}
}

func main() {
	l := [][]int{
		{2, 4, 6, 8},
		{5, 9, 12, 16},
		{1, 11, 5, 9},
		{3, 2, 1, 8},
	}

	m := len(l) // number of rows
	n := len(l) // number of columns
	PrintSpiral(l, m, n)
}

File: code/unclassified/src/spiral_print/spiral_print.java
class Main
{
    private static void printSpiralOrder(int[][] mat)
    {
        // base case
        if (mat == null || mat.length == 0) {
            return;
        }
 
        int top = 0, bottom = mat.length - 1;
        int left = 0, right = mat[0].length - 1;
 
        while (true)
        {
            if (left > right) {
                break;
            }
 
            // print top row
            for (int i = left; i <= right; i++) {
                System.out.print(mat[top][i] + " ");
            }
            top++;
 
            if (top > bottom) {
                break;
            }
 
            // print right column
            for (int i = top; i <= bottom; i++) {
                System.out.print(mat[i][right] + " ");
            }
            right--;
 
            if (left > right) {
                break;
            }
 
            // print bottom row
            for (int i = right; i >= left; i--) {
                System.out.print(mat[bottom][i] + " ");
            }
            bottom--;
 
            if (top > bottom) {
                break;
            }
 
            // print left column
            for (int i = bottom; i >= top; i--) {
                System.out.print(mat[i][left] + " ");
            }
            left++;
        }
    }
 
    public static void main(String[] args)
    {
        int[][] mat =
        {
            { 1, 2, 3, 4, 5},
            {6, 7, 8, 9, 10},
            {11, 12, 13, 14, 15},
            {16, 17, 18, 19, 20},
            {21, 22, 23, 24, 25}
        };
 
        printSpiralOrder(mat);
    }
}